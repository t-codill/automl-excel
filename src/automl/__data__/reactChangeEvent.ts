// tslint:disable: no-any
export const reactChangeEvent: React.ChangeEvent<any> = {
    target: {}
} as unknown as React.ChangeEvent<any>;
