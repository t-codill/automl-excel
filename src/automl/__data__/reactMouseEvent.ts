// tslint:disable-next-line: no-any
export const reactMouseEvent = {} as unknown as React.MouseEvent<any>;
