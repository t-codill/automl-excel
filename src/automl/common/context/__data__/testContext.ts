import { PageNames } from "../../PageNames";
import { Logger } from "../../utils/logger";
import { IWorkspaceContextProps, WorkspaceError } from "../IWorkspaceProps";
import { WorkspaceFlight } from "../WorkspaceFlight";

export function getLogCustomEventSpy(): jest.SpyInstance<ReturnType<Logger["logCustomEvent"]>> {
    const spy = jest.spyOn(Logger.prototype, "logCustomEvent");
    spy.mockImplementation(jest.fn());
    return spy;
}
let pageName: PageNames = PageNames.Unknown;
export const testContext: IWorkspaceContextProps = {
    get pageName(): PageNames { return pageName; },
    get theme(): string { return "light"; },
    discoverUrls: {},
    flight: new WorkspaceFlight(""),
    location: "eastus",
    resourceGroupName: "testResource",
    subscriptionId: "00000000-0000-0000-0000-000000000000",
    logger: new Logger("test"),
    getToken: () => "testToken",
    setPageName: (value: PageNames) => { pageName = value; },
    workspaceName: "testWorkSpace",
    setNavigationBarButtons(): void { return; },
    onError: (err: WorkspaceError) => { throw err; },
    clearErrors: () => { return; },
    setHeader: () => { return; },
    setLoading: () => { return; }
};
