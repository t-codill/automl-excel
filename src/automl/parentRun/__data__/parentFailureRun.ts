import { NotNullableProperties } from "../../common/utils/NotNullableProperties";
import { IParentRunData } from "../ParentRun";

export const parentFailureRun: NotNullableProperties<IParentRunData> = {
    run: {
        runId: "AutoML_000",
        parentRunId: undefined,
        status: "Completed",
        startTimeUtc: new Date("2019-02-01T08:20:00.000Z"),
        endTimeUtc: new Date("2019-02-01T08:30:00.000Z"),
        properties: {
            AMLSettingsJsonString: "",
            primary_metric: "pm",
            errors: "Setup iteration failed: \\n \t\t\t\t\texception: \\n \t\t\t\t\t\tUnforseenConsequence('Invalid subject in the test chamber') \\n"
        }
    },
    experimentName: "Test",
    childRuns: [],
    childRunMetrics: [
        {
            spearman_correlation: 0.877770827248834
        }
    ]
};
