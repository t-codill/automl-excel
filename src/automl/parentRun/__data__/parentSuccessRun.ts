import { NotNullableProperties } from "../../common/utils/NotNullableProperties";
import { IParentRunData } from "../ParentRun";

export const parentSuccessRunAMLSettings = {
  primary_metric: "AUC_weighted",
  data_script: null,
  compute_target: "local",
  n_cross_validations: 3,
  max_cores_per_iteration: 1,
  blacklist_algos: [],
  exit_score: 100,
  name: "my_experiment",
  task_type: "classification"
};
export const parentSuccessRun: NotNullableProperties<IParentRunData> = {
  run: {
    runId: "AutoML_000",
    parentRunId: undefined,
    status: "Completed",
    startTimeUtc: new Date("2019-02-01T08:20:00.000Z"),
    endTimeUtc: new Date("2019-02-01T08:30:00.000Z"),
    properties: {
      AMLSettingsJsonString: JSON.stringify(parentSuccessRunAMLSettings)
    },
    tags: {
      iterations: "2"
    },
  },
  experimentName: "Test",
  childRuns: [
    {
      createdUtc: new Date("2019-02-01T08:20:00.000Z"),
      runId: "AutoML_001",
      parentRunId: "AutoML_000",
      status: "Completed",
      startTimeUtc: new Date("2019-02-01T08:20:00.000Z"),
      endTimeUtc: new Date("2019-02-01T08:25:00.000Z"),
      experimentId: "exp",
      options: {
        generateDataContainerIdIfNotSpecified: true
      },
      name: "AutoML_001",
      properties: {
        runTemplate: "automl_child",
        pipeline_id: "eda663ea8f36e6c07948c99bba9d1becb40991a5",
        run_template: "automl_child",
        run_algorithm: "RandomForest",
        run_preprocessor: "MaxAbsScaler",
        score: "0.977770827248834",
        class_labels: "",
        primary_metric: "spearman_correlation",
      }
    },
    {
      createdUtc: new Date("2019-02-01T08:20:00.000Z"),
      runId: "AutoML_002",
      parentRunId: "AutoML_000",
      status: "Completed",
      startTimeUtc: new Date("2019-02-01T08:20:00.000Z"),
      endTimeUtc: new Date("2019-02-01T08:25:00.000Z"),
      options: {
        generateDataContainerIdIfNotSpecified: true
      },
      name: "AutoML_002",
      properties: {
        runTemplate: "automl_child",
        pipeline_id: "eda663ea8f36e6c07948c99bba9d1becb40991a5",
        run_template: "automl_child",
        run_preprocessor: "MaxAbsScaler",
        score: "0.877770827248834",
        class_labels: "",
        primary_metric: "spearman_correlation",
      }
    },
    {
      createdUtc: new Date("2019-02-01T08:20:00.000Z"),
      runId: "AutoML_003",
      parentRunId: "AutoML_000",
      status: "Canceled",
      startTimeUtc: new Date("2019-02-01T08:20:00.000Z"),
      endTimeUtc: new Date("2019-02-01T08:25:00.000Z"),
      options: {
        generateDataContainerIdIfNotSpecified: true
      },
      name: "AutoML_003",
      properties: {
        runTemplate: "automl_child",
        pipeline_id: "eda663ea8f36e6c07948c99bba9d1becb40991a5",
        run_template: "automl_child",
        score: "0.877770827248834",
        class_labels: "",
        primary_metric: "spearman_correlation",
      }
    },
    {
      createdUtc: new Date("2019-02-01T08:20:00.000Z"),
      runId: "AutoML_004",
      parentRunId: "AutoML_000",
      status: "Completed",
      startTimeUtc: new Date("2019-02-01T08:20:00.000Z"),
      endTimeUtc: new Date("2019-02-01T08:25:00.000Z"),
      options: {
        generateDataContainerIdIfNotSpecified: true
      },
      name: "AutoML_004",
      properties: {
        runTemplate: "automl_child",
        pipeline_id: "eda663ea8f36e6c07948c99bba9d1becb40991a5",
        run_template: "automl_child",
        run_algorithm: "MaxAbsScaler",
        score: "0.827770827248834",
        class_labels: "",
        primary_metric: "spearman_correlation",
      },
    },
    {
      createdUtc: new Date("2019-02-01T08:20:00.000Z"),
      runId: "AutoML_005",
      parentRunId: "AutoML_000",
      status: "Completed",
      startTimeUtc: new Date("2019-02-01T08:20:00.000Z"),
      endTimeUtc: new Date("2019-02-01T08:25:00.000Z"),
      options: {
        generateDataContainerIdIfNotSpecified: true
      },
      name: "AutoML_005",
    }
  ],
  childRunMetrics: [
    {
      spearman_correlation: 0.877770827248834
    }
  ]
};
