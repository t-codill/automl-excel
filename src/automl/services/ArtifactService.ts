import { ArtifactAPI, ArtifactAPIModels } from "@vienna/artifact";
import { blob2string } from "../common/utils/blob2string";
import { IServiceBaseProps } from "./ServiceBase";
import { ServiceBaseNonArm } from "./ServiceBaseNonArm";

const artifactClientApiVersion = "1.0";
const runOrigin = "ExperimentRun";
const modelPath = "outputs/model.pkl";

export interface IArtifact {
    origin: string;
    container: string;
    path: string;
}

export interface IRunDataForArtifact {
    runId: string | undefined;
    parentRunId: string | undefined;
    status: string | undefined;
}

async function getArtifactContent(artifact: ArtifactAPIModels.GetArtifactContentByIdResponse): Promise<string | undefined> {
    const blobBody = await artifact.blobBody;
    if (blobBody) {
        return blob2string(blobBody);
    }
    return undefined;
}
export class ArtifactService extends ServiceBaseNonArm<ArtifactAPI> {
    constructor(props: IServiceBaseProps) {
        super(props, ArtifactAPI, props.discoverUrls.history);
    }
    public async tryGetContentForRun(runId: string, path: string): Promise<string | null | undefined> {
        return this.tryGetContent({ origin: runOrigin, container: runId, path });
    }

    public async tryGetContent(artifact: IArtifact): Promise<string | null | undefined> {
        return this.trySend(async (client, abortSignal) => {
            const artifactContent = await client.getArtifactContentById2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                artifact.origin,
                artifact.container,
                artifactClientApiVersion, {
                    path: artifact.path,
                    abortSignal
                });
            return getArtifactContent(artifactContent);
        });
    }

    public async getContent(artifact: IArtifact): Promise<string | undefined> {
        return this.send(async (client, abortSignal) => {
            const artifactContent = await client.getArtifactContentById2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                artifact.origin,
                artifact.container,
                artifactClientApiVersion, {
                    path: artifact.path,
                    abortSignal
                });
            return getArtifactContent(artifactContent);
        });
    }

    public async getAllContents(params: IArtifact[]): Promise<Array<string | undefined> | undefined> {
        return this.parallelGetAllValues(params, async (param, client, abortSignal) => {
            const artifact = await client.getArtifactContentById2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                param.origin,
                param.container,
                artifactClientApiVersion, {
                    path: param.path,
                    abortSignal
                });
            return getArtifactContent(artifact);
        });
    }

    public async getAllArtifactsForRuns(runs: string[]): Promise<ArtifactAPIModels.ArtifactDto[][] | undefined> {
        return this.parallelGetAllValuesWithContinuationToken(runs, async (runId, client, abortSignal, continuationToken) => {
            return client.getArtifactsInContainerOrPath2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                runOrigin,
                runId,
                artifactClientApiVersion, {
                    abortSignal,
                    continuationToken
                });
        });
    }

    public async getAllArtifacts(runId: string): Promise<ArtifactAPIModels.ArtifactDto[] | undefined> {
        return this.getAllValuesWithContinuationToken(async (client, abortSignal, continuationToken) => {
            return client.getArtifactsInContainerOrPath2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                runOrigin,
                runId,
                artifactClientApiVersion, {
                    abortSignal,
                    continuationToken
                });
        });
    }

    public async tryGetArtifactUrl(artifact: IArtifact): Promise<ArtifactAPIModels.GetArtifactContentInformation2Response | null | undefined> {
        return this.trySend(async (client, abortSignal) => {
            return client.getArtifactContentInformation2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                artifact.origin,
                artifact.container,
                artifactClientApiVersion, {
                    abortSignal,
                    path: artifact.path
                });
        });
    }

    public async getModelUrl(run: IRunDataForArtifact): Promise<string | null | undefined> {
        if (!run || !run.runId || !run.parentRunId || run.status !== "Completed") {
            return null;
        }
        const result = await this.tryGetArtifactUrl({ container: run.runId, origin: runOrigin, path: modelPath });
        if (result === undefined) {
            return undefined;
        }
        if (!result || !result.contentUri) {
            return null;
        }
        return result.contentUri;
    }
}
