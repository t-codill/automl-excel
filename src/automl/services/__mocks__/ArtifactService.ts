import { ArtifactAPIModels } from "@vienna/artifact";

export interface IArtifact {
    origin: string;
    container: string;
    path: string;
}

export class ArtifactService {
    private readonly tryGetContentResult = "tryGetContent";
    private readonly getAllArtifactsResult: ArtifactAPIModels.ArtifactDto[] = [
        {
            artifactId: "artifactId1",
            container: "container1",
            origin: "origin1",
            path: "path1"
        },
        {
            artifactId: "artifactId2",
            container: "container2",
            origin: "origin2",
            path: "path2"
        }
    ];
    private readonly tryGetArtifactUrlResult: ArtifactAPIModels.ArtifactContentInformationDto = {
        contentUri: "contentUri",
        path: "path"
    };
    private readonly getModelUrlResult = "getModelUrl";
    private readonly getModelUrlsResult = ["getModelUrls1", "getModelUrls2"];

    public async tryGetContentForRun(): Promise<string | null | undefined> {
        return this.tryGetContentResult;
    }

    public async tryGetContent(): Promise<string | null | undefined> {
        return this.tryGetContentResult;
    }

    public async getContent(): Promise<string | null | undefined> {
        return this.tryGetContentResult;
    }

    public async getAllContents(params: IArtifact[]): Promise<Array<string | undefined> | undefined> {
        return params.map((p) => `"getAllContentsResult:${p.origin}/${p.container}/${p.path}`);
    }

    public async getAllArtifacts(): Promise<ArtifactAPIModels.ArtifactDto[] | undefined> {
        return this.getAllArtifactsResult;
    }

    public async getAllArtifactsForRuns(runs: string[]): Promise<ArtifactAPIModels.ArtifactDto[][] | undefined> {
        return runs.map(() => this.getAllArtifactsResult);
    }

    public async tryGetArtifactUrl(): Promise<ArtifactAPIModels.ArtifactContentInformationDto | null | undefined> {
        return this.tryGetArtifactUrlResult;
    }

    public async getModelUrl(): Promise<string | undefined> {
        return this.getModelUrlResult;
    }

    public async getModelUrls(): Promise<Array<string | null> | undefined> {
        return this.getModelUrlsResult;
    }

    public dispose(): void {
        return;
    }

    public reset(): void {
        return;
    }
}
