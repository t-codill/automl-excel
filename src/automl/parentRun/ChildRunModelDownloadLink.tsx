import { Icon, Link, Spinner, SpinnerSize } from "office-ui-fabric-react";
import * as React from "react";
import { download } from "../common/utils/download";
import { BaseComponent } from "../components/Base/BaseComponent";
import { ArtifactService, IRunDataForArtifact } from "../services/ArtifactService";

interface IChildRunModelDownloadLinkProps {
    run: IRunDataForArtifact;
}

export class ChildRunModelDownloadLink extends BaseComponent<
    IChildRunModelDownloadLinkProps,
    {
        downloading: boolean;
    },
    { artifactService: ArtifactService }
    > {
    protected readonly serviceConstructors = { artifactService: ArtifactService };
    protected getData?: (() => Promise<void>) | undefined;
    constructor(props: IChildRunModelDownloadLinkProps) {
        super(props);
        this.state = {
            downloading: false
        };
    }
    public render(): React.ReactNode {
        if (this.props.run.status !== "Completed") {
            return <></>;
        }
        if (this.state.downloading) {
            return <Spinner size={SpinnerSize.small} />;
        }

        return <Link onClick={this.onClick}>
            <Icon iconName="Download" /> Download
        </Link>;
    }

    private readonly onClick = async () => {
        this.setState({ downloading: true });
        const uri = await this.services.artifactService.getModelUrl(this.props.run);
        if (uri === undefined) {
            return;
        }
        this.setState({ downloading: false });
        if (uri === null) {
            return;
        }
        this.logUserAction("ModelDownloadFromTable", { childRunId: this.props.run.runId });
        download(uri, "model.pkl");
    }
}
