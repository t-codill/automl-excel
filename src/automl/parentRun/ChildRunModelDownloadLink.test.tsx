import { shallow } from "enzyme";
import { Link } from "office-ui-fabric-react";
import * as React from "react";
import { reactMouseEvent } from "../__data__/reactMouseEvent";
import { getLogCustomEventSpy, testContext } from "../common/context/__data__/testContext";
import { ArtifactService } from "../services/ArtifactService";
import { ChildRunModelDownloadLink } from "./ChildRunModelDownloadLink";

jest.mock("../services/ArtifactService");
let asGetModelUrl: jest.SpyInstance<ReturnType<ArtifactService["getModelUrl"]>>;

describe("Child Run Model Download Link", () => {
    const completedRunData = {
        runId: "001",
        parentRunId: "000",
        status: "Completed"
    };
    const failedRunData = {
        runId: "001",
        parentRunId: "000",
        status: "Failed"
    };
    beforeEach(() => {
        asGetModelUrl = jest.spyOn(ArtifactService.prototype, "getModelUrl");
    });
    describe("completed run", () => {
        const props = {
            run: completedRunData,
            downloading: false
        };

        it("should render link when downloading is false", () => {

            const tree = shallow(<ChildRunModelDownloadLink {...props} />);
            expect(tree)
                .toMatchSnapshot();
        });
        it("should download when model uri is defined", async () => {
            const logSpy = getLogCustomEventSpy();
            const mockGetUrl = Promise.resolve("mockUri");
            asGetModelUrl.mockReturnValueOnce(mockGetUrl);
            const c = shallow(<ChildRunModelDownloadLink {...props} />);
            const link = c.find(Link)
                .props();
            if (link && link.onClick) {
                link.onClick(reactMouseEvent);
            }
            await Promise.resolve();
            const downloading: boolean = c.state("downloading");
            expect(downloading)
                .toBe(false);
            expect(logSpy)
                .toBeCalledWith("_ModelDownloadFromTable_UserAction",
                    testContext,
                    { childRunId: "001", component: "ModelDownloadFromTable", pageName: "" });
        });
        it("should not download when model uri is undefined", async (done) => {
            const mockGetUrl = Promise.resolve(undefined);
            asGetModelUrl.mockReturnValueOnce(mockGetUrl);
            const tree = shallow(<ChildRunModelDownloadLink {...props} />);
            await mockGetUrl;
            const link = tree.find(Link)
                .props();
            if (link && link.onClick) {
                link.onClick(reactMouseEvent);
                setImmediate(() => {
                    const downloading: boolean = tree.state("downloading");
                    expect(downloading)
                        .toBe(true);
                    done();
                });
            }
        });
        it("should not download when model uri is null", async (done) => {
            const mockGetUrl = Promise.resolve(null);
            asGetModelUrl.mockReturnValueOnce(mockGetUrl);
            const tree = shallow(<ChildRunModelDownloadLink {...props} />);
            await mockGetUrl;
            const link = tree.find(Link)
                .props();
            if (link && link.onClick) {
                link.onClick(reactMouseEvent);
                setImmediate(() => {
                    const downloading: boolean = tree.state("downloading");
                    expect(downloading)
                        .toBe(false);
                    done();
                });
            }
        });
        it("should render spinner when downloading is true", () => {
            props.downloading = true;
            const tree = shallow(<ChildRunModelDownloadLink {...props} />);
            expect(tree)
                .toMatchSnapshot();
        });
    });
    it("should not render anything when run failed", () => {
        const props = {
            run: failedRunData,
            downloading: true
        };
        const tree = shallow(<ChildRunModelDownloadLink {...props} />);
        expect(tree)
            .toMatchSnapshot();
    });
});
