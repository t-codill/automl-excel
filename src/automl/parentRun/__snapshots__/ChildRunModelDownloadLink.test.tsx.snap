// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Child Run Model Download Link completed run should render link when downloading is false 1`] = `
<StyledLinkBase
  onClick={[Function]}
>
  <StyledIconBase
    iconName="Download"
  />
   Download
</StyledLinkBase>
`;

exports[`Child Run Model Download Link completed run should render spinner when downloading is true 1`] = `
<StyledLinkBase
  onClick={[Function]}
>
  <StyledIconBase
    iconName="Download"
  />
   Download
</StyledLinkBase>
`;

exports[`Child Run Model Download Link should not render anything when run failed 1`] = `<Fragment />`;
